Maple2.Server.Game/PacketHandlers/QuestHandler.cs

```diff
@@ -0,0 +1,125 @@
+        ResumeDungeon = 19,
+        Dispatch = 20,
+        RemoteComplete = 24, // Maybe? This is mainly used for Navigator
+    }
+
+    public override void Handle(GameSession session, IByteReader packet) {
+        var command = packet.Read<Command>();
+        switch (command) {
+            case Command.Accept:
+                HandleAccept(session, packet);
+                break;
+            case Command.Complete:
+                HandleComplete(session, packet);
+                break;
+            case Command.Abandon:
+                HandleForfeit(session, packet);
+                break;
+            case Command.Tracking:
+                HandleSetTracking(session, packet);
+                break;
+            case Command.GoToNpc:
+                HandleGoToNpc(session, packet);
+                break;
+            case Command.RemoteComplete:
+                HandleRemoteComplete(session, packet);
+                return;
+        }
+    }
+
+    private static void HandleAccept(GameSession session, IByteReader packet) {
+        int questId = packet.ReadInt();
+        int npcObjectId = packet.ReadInt();
+
+        if (session.Field == null || !session.Field.Npcs.TryGetValue(npcObjectId, out FieldNpc? npc)) {
+            return; // Invalid Npc
+        }
+
+        session.Quest.Start(questId);
+    }
+
+    private void HandleComplete(GameSession session, IByteReader packet) {
+        int questId = packet.ReadInt();
+
+        if (!session.Quest.TryGetQuest(questId, out Quest? quest)) {
+            return;
+        }
+        if (!session.Quest.Complete(quest)) {
+            Logger.Warning("Could not complete quest {QuestId}", questId);
+        }
+    }
+
+    private static void HandleForfeit(GameSession session, IByteReader packet) {
+        int questId = packet.ReadInt();
+
+        if (!session.Quest.TryGetQuest(questId, out Quest? quest) || !quest.Metadata.Basic.Forfeitable) {
+            return;
+        }
+
+        if (quest.CompletionCount > 0) {
+            quest.State = QuestState.Completed; // ?? how do you revert?
+        }
+
+        if (session.Quest.Remove(quest)) {
+            session.Send(QuestPacket.Abandon(quest.Id));
+        }
+    }
+
+    private static void HandleSetTracking(GameSession session, IByteReader packet) {
+        int questId = packet.ReadInt();
+        bool tracking = packet.ReadBool();
+
+        if (!session.Quest.TryGetQuest(questId, out Quest? quest) || quest.State is QuestState.Completed) {
+            return;
+        }
+
+        session.Send(QuestPacket.SetTracking(questId, tracking));
+    }
+
+    private static void HandleGoToNpc(GameSession session, IByteReader packet) {
+        int questId = packet.ReadInt();
+
+        if (!session.Quest.TryGetQuest(questId, out Quest? quest) || !quest.Metadata.GoToNpc.Enabled) {
+            return;
+        }
+
+        session.Send(session.PrepareField(quest.Metadata.GoToNpc.MapId, portalId: quest.Metadata.GoToNpc.PortalId)
```

Comment:
is there a check if the current map allows warping?

Reply:
I do recall there being some check in game but I'm not sure what is the value in the xml that determines it

Reply:
I was thinking it might be `Move`: https://github.com/kOchirasu/Maple2/blob/8161c46fc6020f9a4560226be92f3796f4cc768c/Maple2.Model/Metadata/MapMetadata.cs#L76